src/tree/updater_gpu_hist.cu

```diff
@@ -1339,35 +1328,36 @@ inline void DeviceShard<GradientSumT>::CreateHistIndices(
     }
     size_t batch_nrows = batch_row_end - batch_row_begin;
 
+    /*! \brief row offset in SparsePage (the input data). */
+    dh::device_vector<size_t> row_ptrs(batch_nrows+1);
+    thrust::copy(
+      offset_vec.data() + device_row_state.row_offset_in_current_batch + batch_row_begin,
+      offset_vec.data() + device_row_state.row_offset_in_current_batch + batch_row_end + 1,
+      row_ptrs.begin());
+
     // number of entries in this batch.
-    size_t n_entries = row_ptrs[batch_row_end] - row_ptrs[batch_row_begin];
+    size_t n_entries = row_ptrs.back() - row_ptrs.front();
```

Comment:
I never noticed this before, but isn't the dereferencing elements of `row_ptrs` on the host invoking a bunch of tiny memory copies from the device to host?

Reply:
yes, it will. i have made the requisite changes; please take 1 more look.

fwiw, there is no perceptible change in performance with this.